### PR TITLE
Feature/memento strategy

### PR DESCRIPTION
**Pull request by CaptainShayfish**

### CODE REVIEW

**Review by RaiderRoss — APPROVED** [review]

File: src/policies/base/Policy.h [hunk 1]
```diff
@@ -1,11 +1,55 @@
 #ifndef POLICY_H
 #define POLICY_H
 
-class Policy
-{
+#include "Memento.h"
```
Comment (jacqu3sk) [inline src/policies/base/Policy.h:4]: Hey there is an import issue here can you just make sure to import "utils/Memento.h" not "Memento.h"

File: tests/utils/MementoTest.cpp [hunk 2]
```diff
@@ -1,6 +1,28 @@
+#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
 #include "doctest.h"
+#include "Memento.h"
```
Comment (jacqu3sk) [inline tests/utils/MementoTest.cpp:3]: Also please fix the import here it should be `#include "utils/Memento.h"` not `#include "Memento.h"`

File: tests/utils/CaretakerTest.cpp [hunk 3]
```diff
@@ -0,0 +1,27 @@
+#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
+#include "doctest.h"
+#include "Caretaker.h"
+#include "Memento.h"
```
Comment (jacqu3sk) [inline tests/utils/CaretakerTest.cpp:4]: also fix the includes here to `#include "Memento.h"` should be `#include "utils/Memento.h"` and `#include "Caretaker.h"` should be `#include "utils/Caretaker.h"`

File: tests/policies/water/LowWaterPolicy.cpp [hunk 4]
```diff
@@ -1,6 +1,14 @@
+#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
 #include "doctest.h"
+#include "LowWaterPolicy.h"
```
Comment (jacqu3sk) [inline tests/policies/water/LowWaterPolicy.cpp:3]: Also fix the import here it should be #include "policies/water/LowWaterPolicy.h" and not #include "LowWaterPolicy.h"

**Review by jacqu3sk — CHANGES_REQUESTED** [review]
Please fix all the imports I have noted

File: tests/policies/base/PolicyTest.cpp [hunk 5]
```diff
@@ -1,6 +1,26 @@
+#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
```
Comment (jacqu3sk) [inline tests/policies/base/PolicyTest.cpp:1]: Remove this line please

File: tests/policies/electricity/HighElectricityPolicyTest.cpp [hunk 6]
```diff
@@ -1,6 +1,14 @@
+#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
```
Comment (jacqu3sk) [inline tests/policies/electricity/HighElectricityPolicyTest.cpp:1]: Remove this line please

File: tests/policies/electricity/LowElectricityPolicyTest.cpp [hunk 7]
```diff
@@ -1,6 +1,14 @@
+#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
```
Comment (jacqu3sk) [inline tests/policies/electricity/LowElectricityPolicyTest.cpp:1]: Remove this line please

File: tests/policies/electricity/NormalElectricityPolicyTest.cpp [hunk 8]
```diff
@@ -0,0 +1,13 @@
+#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
```
Comment (jacqu3sk) [inline tests/policies/electricity/NormalElectricityPolicyTest.cpp:1]: Remove this line please

File: tests/policies/water/HighWaterPolicy.cpp [hunk 9]
```diff
@@ -1,6 +1,14 @@
+#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
```
Comment (jacqu3sk) [inline tests/policies/water/HighWaterPolicy.cpp:1]: Remove this line please

File: tests/policies/water/LowWaterPolicy.cpp [hunk 10]
```diff
@@ -1,6 +1,14 @@
+#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
```
Comment (jacqu3sk) [inline tests/policies/water/LowWaterPolicy.cpp:1]: Remove this line please

File: tests/policies/water/NormalWaterPolicy.cpp [hunk 11]
```diff
@@ -0,0 +1,13 @@
+#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
```
Comment (jacqu3sk) [inline tests/policies/water/NormalWaterPolicy.cpp:1]: Remove this line please

File: tests/utils/CaretakerTest.cpp [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+
+    // Retrieve the most recent memento (m2)
+    Memento* retrieved = caretaker.getMemento();
+    
+    CHECK(retrieved->getName() == "Policy2");
+    CHECK(retrieved->getDetail() == "Detail2");
+
+    // Clean up memory
+    delete m1;
+    delete m2;
```
Comment (jacqu3sk) [inline tests/utils/CaretakerTest.cpp:26]: Also remove these two delete lines they cause a double free error

**Review by jacqu3sk — CHANGES_REQUESTED** [review]
Please view all the comments I made and make the changes please

**Review by jacqu3sk — CHANGES_REQUESTED** [review]

**Review by jacqu3sk — APPROVED** [review]
Everything looks good now

**Review by RaiderRoss — APPROVED** [review]

**Pull request merged**